Holovibes/cuda_sources/otsu.cu

```diff
@@ -46,98 +50,139 @@ __global__ void bradley_threshold_kernel(const float* image,
                                          float globalThreshold,
                                          float localThresholdFactor)
 {
-    int idx = blockIdx.x * blockDim.x + threadIdx.x;
-
-    int x = idx % width;
-    int y = idx / width;
-
-    if (x >= width || y >= height)
-        return;
+    int x = blockIdx.x * blockDim.x + threadIdx.x;
+    int y = blockIdx.y * blockDim.y + threadIdx.y;
 
-    int halfWindow = windowSize / 2;
-    int startX = max(x - halfWindow, 0);
-    int startY = max(y - halfWindow, 0);
-    int endX = min(x + halfWindow, width - 1);
-    int endY = min(y + halfWindow, height - 1);
+    if (x < width || y < height)
+    {
+        int halfWindow = windowSize / 2;
+        int startX = max(x - halfWindow, 0);
+        int startY = max(y - halfWindow, 0);
+        int endX = min(x + halfWindow, width - 1);
+        int endY = min(y + halfWindow, height - 1);
 
-    float localSum = 0;
-    int count = 0;
+        float localSum = 0;
+        int count = 0;
 
-    for (int i = startX; i <= endX; i++)
-    {
-        for (int j = startY; j <= endY; j++)
+        for (int i = startX; i <= endX; i++)
         {
-            localSum += image[j * width + i];
-            count++;
+            for (int j = startY; j <= endY; j++)
+            {
+                localSum += image[j * width + i];
+                count++;
+            }
         }
-    }
 
-    float localMean = localSum / count;
-    float localThreshold = localMean * (1 - localThresholdFactor * globalThreshold);
-    output[y * width + x] = (image[y * width + x] > localThreshold) ? 1.0f : 0.0f;
+        float localMean = localSum / count;
+        float localThreshold = localMean * (1 - localThresholdFactor * globalThreshold);
+        output[y * width + x] = (image[y * width + x] > localThreshold) ? 1.0f : 0.0f;
```

Comment:
Maybe use some arithmetic operators.

Reply:
For this function in the future have a look to add some GPU side reduction.

Reply:
yep